### DISCUSSION TITLE
Channel EQ

### DISCUSSION THREAD

**Discussion by RafaSRibeiro** [kryops/remote-mixer#2] [Ideas]
Hello
I'm interested in contributing to the development of the EQ feature.
Any tips on how to get started?

**Comment by kryops** [comment]
Hi @RafaSRibeiro,

that sounds awesome!
I don't have much free time (as usual), but I can try and support you if you have questions or run into issues.

Most features roughly consist of the following parts:
* New properties in the `DeviceConfiguration` to enable and configure the feature for different mixing console models
* New properties in the `RemoteMixerState` to keep the current state of the EQ and sync it to the frontend
* Maybe a new type of `ApiInMessage`/`ApiOutMessage` for features that do more than setting a specific property per channel (which would be handled via `ApiChangeMessage`)
* The implementation in the `DeviceController` for the mixer types that support the feature
* Some UI for it in the frontend

A word of warning: Channel EQs probably aren't the easiest feature for a first contribution - it was probably more than lack of time that I haven't tried to implement them myself 😅 

Here are some of my thoughts:

* As it would be great for them to work across different mixers, the first challenge is probably to define a config/state format that is flexible enough to do that: I'd start with something like "each channel can have n (usually 4) EQ bands that have a following properties (frequency, gain, value?)". The second step would be to account for different EQ types especially for the low and high bands (e.g. high/low pass or shelf)
* Building a good UI for a parametric EQ is quite a task on its own - just drawing a line for the current settings sounds like a whole lot of maths combined with something like SVG or a Canvas for rendering. Maybe you can start off with a few faders and plain text values and iterate from there 🤔 

That said, good luck and have fun, and just message if you have questions or issues - though I might take a few days to reply 🙈
